### PR TITLE
Backend: Add multi version preprocessor

### PR DESCRIPTION
**Pull request by lineargraph**
## PR Reviews

When your PR is marked as ready for review, some of our maintainers will look through your code to make sure everything is good to go. In order to do this, they may request some changes you will need to do, **or fix smaller stuff (like merge conflicts) for you**. If a maintainer has reviewed your PR, make sure to **pull any of their changes** into your local project before doing more work on your code. Having maintainers fix small stuff for you helps us speed up the process of merging your PR, so if some of your systems warrant further care, be sure to let us know (preferably with a code comment).

Make sure to only mark your PR as "Ready to review" when it is. If you still want to do major changes, you can keep a draft PR open until then.

## What
This PR adds a preprocessor to automatically remap all skyhanni code to 1.21. This does not work, of course. The remapped sources are full of problems: Some methods have different names whose conversion gets lost between versions. Some methods simply get called with different arguments and need to manually handled. Some methods get fully removed and new workarounds need to be written.

The goal of this PR is to have a way for contributors to slowly make features they write work on 1.21 over time to ease the burden when making the final jump. It should run in the background and not halt normal development.

Check out the CONTRIBUTING.md changes for more info.

## Changelog Technical Details
+ Added a preprocessor to prepare for 1.21. - !nea
    * This will require a minor adjustment in how the Minecraft client is launched.

### CODE REVIEW

**Review by pauliesnug — APPROVED** [review]

**Comment by lineargraph** [comment]
I am going to assume that the conflict here is the version number only and am not going to fix the conflict therefore. Unless you wanna make bigger changes you should probably just discard the changes and edit the version number directly in the new branch

**Comment by lineargraph** [comment]
i didnt mean for you to fix it i just meant that i didnt feel like fixing it for every version update lol

**Comment by hannibal002** [comment]
I intend to merge this pr before the next beta version

**Comment by btwonion** [comment]
Hi,
in the Contributing document you described the processing of the mappings very extensively.
But how do you plan to replace the forge libraries? The main problem here is the event system I suppose - are you planning to replace it completely for a standalone system?

**Comment by lineargraph** [comment]
For the event library there is already a replacement system existing that is planned to be used in 1.8.9 as well. For the other forge functions we'll need to write custom functions as well using preprocessor directives like for other missing minecraft functions.

**Comment by btwonion** [comment]
ah I see. great. I'm looking forward to the porting process.

**Comment by hannibal002** [comment]
this reminds me, we need to readdress the event bus system and finally move over to the new system globally (after fixed all remaining problems with it)

**Review by hannibal002 — CHANGES_REQUESTED** [review]
After running this in dev env, ive found two bugs:

They happened when running the 1.8.9 runClient script. I am not sure if this is related, though.
image



<details>
<summary>shaders are broken in custom scoreboard</summary>

```
SkyHanni 0.26.Beta.21: Caught an ExceptionInInitializerError in CustomScoreboard at GuiOverlayRenderEvent: null
 
Caused by java.lang.ExceptionInInitializerError: null
at SH.utils.RenderUtils.drawRoundRect(RenderUtils.kt:1785)
at SH.utils.RenderUtils.drawRoundRect$default(RenderUtils.kt:1778)
at SH.features.gui.customscoreboard.RenderBackground.renderBackground(RenderBackground.kt:60)
at SH.features.gui.customscoreboard.CustomScoreboard.onRenderOverlay(CustomScoreboard.kt:61)
at SH.data.RenderData.onRenderOverlay(RenderData.kt:31)
at FML.common.eventhandler.EventBus.post(EventBus.java:140)
Caused by java.io.FileNotFoundException: skyhanni:shaders/rounded_rect.vsh
at MC.client.resources.SimpleReloadableResourceManager.getResource(SimpleReloadableResourceManager.java:68)
at SH.utils.shader.ShaderManager.loadShader(ShaderManager.kt:67)
at SH.utils.shader.Shader.recompile(Shader.kt:59)
at SH.utils.shader.Shader.<init>(Shader.kt:31)
at SH.features.misc.RoundedShader.<init>(RoundedShader.kt:7)
at SH.features.misc.RoundedRectangleShader.<init>(RoundedShader.kt:27)
at SH.features.misc.RoundedRectangleShader.<clinit>(RoundedShader.kt)
at SH.utils.RenderUtils.drawRoundRect(RenderUtils.kt:1785)
at SH.utils.RenderUtils.drawRoundRect$default(RenderUtils.kt:1778)
at SH.features.gui.customscoreboard.RenderBackground.renderBackground(RenderBackground.kt:60)
at SH.features.gui.customscoreboard.CustomScoreboard.onRenderOverlay(CustomScoreboard.kt:61)
at SH.data.RenderData.onRenderOverlay(RenderData.kt:31)
at FML.common.eventhandler.EventBus.post(EventBus.java:140)
```

</details>

<details>
<summary>Skull texture gets rendered weirdly wrong</summary>

image

</details>

**Comment by ItsEmpa** [comment]
the skull being rendered wrong is because of the same shaders bug

**Comment by lineargraph** [comment]
Looks like it's just forge not finding the resources because they are not in the kotlin classes directory. Forgot it is like that on 1.8.9. As for the runClient: the runClient gradle task should work but I'll also make it generate a regular intellij config again (Although it should be noted that the intellij task does not seem manual set up to work with the right java version, unlike the gradle task)

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-8.3-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-8.7-bin.zip
```
Comment (pauliesnug) [inline gradle/wrapper/gradle-wrapper.properties]: bump gradle version to `8.9` question mark

File: settings.gradle.kts [hunk 2]
```diff
@@ -20,8 +30,20 @@ pluginManagement {
 }
 
 plugins {
-    id("org.gradle.toolchains.foojay-resolver-convention") version("0.6.0")
+    id("org.gradle.toolchains.foojay-resolver-convention") version ("0.6.0")
```
Comment (pauliesnug) [inline settings.gradle.kts]: bump to 0.8.0/use libs.versions.toml

File: settings.gradle.kts [hunk 3]
```diff
@@ -9,6 +13,12 @@ pluginManagement {
         maven("https://repo.spongepowered.org/maven/")
         maven("https://repo.nea.moe/releases")
         maven("https://repo.sk1er.club/repository/maven-releases/")
+        maven("https://maven.deftu.xyz/releases")
+        maven("https://jitpack.io") {
+            content {
+                includeGroupByRegex("(com|io)\\.github\\..*")
```
Comment (pauliesnug) [inline settings.gradle.kts:19]: is this not just all of jitpack?

File: sharedVariables/src/MinecraftVersion.kt [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    val versionName: String,
+    val javaVersion: Int,
+) {
+    MC189("1.8.9", 8),
+    MC11200("1.12", 8),
+    MC11202("1.12.2", 8),
+    MC1144("1.14.4", 8),
+    MC11605("1.16.5", 8),
+    MC121("1.21", 21),
+    ;
```
Comment (pauliesnug) [inline sharedVariables/src/MinecraftVersion.kt:15]: is having the enum delimiter `;` be dangling convention /genq

**Review by pauliesnug — CHANGES_REQUESTED** [review]
looks gay to me!!!!!! :3 /pos

File: sharedVariables/src/MinecraftVersion.kt (hunk 4 above)
Comment (pauliesnug) [reply sharedVariables/src/MinecraftVersion.kt:15]: does skyhanni have a standard formatting for kotlin; if not might be better discussion in an issue/new pr

Comment (hannibal002) [reply sharedVariables/src/MinecraftVersion.kt:15]: Skyhanni has a lint setting that your ide can use. especially here, we explicitly want to show trialing commas always.

File: settings.gradle.kts (hunk 3 above)
Comment (lineargraph) [reply settings.gradle.kts:19]: It makes other artifact coordinates skip that repo

File: sharedVariables/src/MinecraftVersion.kt (hunk 4 above)
Comment (lineargraph) [reply sharedVariables/src/MinecraftVersion.kt:15]: > is having the enum delimiter ; be dangling convention /genq

i dont know if that is a convention for skyhanni or kotlin in general, but i prefer having all my enums declared with a comma and a semicolon on the next line so i can comment / uncomment entries with a line comment without breaking anything. same goes for argument lists

File: settings.gradle.kts (hunk 3 above)
Comment (pauliesnug) [reply settings.gradle.kts:19]: ah okay

**Pull request merged**